flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -76,86 +77,163 @@ class GradientDescent(runParameters: ParameterMap) extends IterativeSolver {
+          iterateWithTermination(numberOfIterations) {
+          weightsWithLossSum =>
+
+            // Extract weight vector and loss
+            val previousWeightsDS = weightsWithLossSum.map{_._1}
+            val previousLossSumDS = weightsWithLossSum.map{_._2}
+
+            val currentWeightsDS = SGDStep(data, previousWeightsDS)
+
+            val currentLossSumDS = lossCalculation(data, currentWeightsDS)
+
+            // Check if the relative change in the loss is smaller than the
+            // convergence threshold. If yes, then terminate i.e. return empty termination data set
+            val termination = previousLossSumDS.crossWithTiny(currentLossSumDS).setParallelism(1).
+              filter{
+              pair => {
+                val (previousLoss, currentLoss) = pair
+
+                if (previousLoss <= 0) {
+                  false
+                } else {
+                  math.abs((previousLoss - currentLoss)/previousLoss) >= convergence
+                }
+              }
+            }
+
+            // Result for new iteration
+            (currentWeightsDS cross currentLossSumDS, termination)
+        }
+        // Return just the weights
+        resultWithLoss.map{_._1}
     }
+    optimizedWeights
   }
 
-  /** Mapping function that calculates the weight gradients from the data.
+  /** Calculates the loss value, given a labeled vector and the current weight vector
     *
+    * The weight vector is received as a broadcast variable.
     */
-  private class GradientCalculation extends
-    RichMapFunction[LabeledVector, (WeightVector, Double, Int)] {
+  private class LossCalculation extends RichMapFunction[LabeledVector, (Double, Int)] {
 
     var weightVector: WeightVector = null
 
+
     @throws(classOf[Exception])
     override def open(configuration: Configuration): Unit = {
       val list = this.getRuntimeContext.
         getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
 
       weightVector = list.get(0)
-    }
 
-    override def map(example: LabeledVector): (WeightVector, Double, Int) = {
+    }
 
-      val lossFunction = parameterMap(LossFunction)
-      val regType = parameterMap(RegularizationType)
-      val regParameter = parameterMap(RegularizationParameter)
-      val predictionFunction = parameterMap(PredictionFunctionParameter)
+    override def map(example: LabeledVector): (Double, Int) = {
+      val lossFunction = parameters(LossFunctionParameter)
+      val predictionFunction = parameters(PredictionFunctionParameter)
       val dimensions = example.vector.size
-      // TODO(tvas): Any point in carrying the weightGradient vector for in-place replacement?
-      // The idea in spark is to avoid object creation, but here we have to do it anyway
+      // TODO(tvas): Avoid needless creation of WeightGradient object
+      // Create a lossValue function in LossFunction?
       val weightGradient = new DenseVector(new Array[Double](dimensions))
```

Comment:
I would probably like to change this. Worth it to add another public function to LossFunction?